sale_loyalty_auto_refresh/models/sale_order.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2021 Tecnativa - David Vidal
+# Copyright 2021 Camptocamp - Silvio Gregorini
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+import logging
+
+from odoo import api, fields, models
+
+logger = logging.getLogger(__name__)
+
+
+class SaleOrder(models.Model):
+    _name = "sale.order"
+    _inherit = ["sale.order", "sale.coupon.refresh.mixin"]
+
+    # Used in UI to hide the manual button
+    auto_refresh_coupon = fields.Boolean(
+        related="company_id.auto_refresh_coupon",
+    )
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        if self._check_skip_refresh():
+            return super().create(vals_list)
+
+        self_ctx = self.with_context(skip_auto_refresh_coupons=True)
+        orders = super(SaleOrder, self_ctx).create(vals_list)
+        orders._auto_refresh_coupons()
+        return orders
+
+    def write(self, vals):
+        if self._check_skip_refresh():
+            return super().write(vals)
+
+        old_data = self._read_recs_data()
+        self_ctx = self.with_context(skip_auto_refresh_coupons=True)
+        res = super(SaleOrder, self_ctx).write(vals)
+        new_data = self._read_recs_data()
+        # Until we restart Odoo, we won't get new triggers from params. Once restarted
+        # the method will return an empty set.
+        new_triggers = self._new_trigger()
+        if old_data != new_data or any(x in new_triggers for x in vals):
+            self._auto_refresh_coupons()
+        return res
+
+    def _auto_refresh_coupons(self):
+        self_ctx = self.with_context(skip_auto_refresh_coupons=True)
+        for order in self_ctx:
+            if order._allow_recompute_coupon_lines():
+                order._update_programs_and_rewards()
+                order.action_apply_rewards()
+
+    def action_apply_rewards(self):
+        self.ensure_one()
+        claimable_rewards = self._get_claimable_rewards()
+        for coupon, reward in claimable_rewards.items():
```

Comment:
This part is for [F#63041](https://tms.trobz.com/web#id=63041&action=543&model=tms.forge.ticket&view_type=form&cids=1&menu_id=424)
- Get all the programs and try to apply them.